superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/utils.ts

```diff
@@ -25,40 +25,48 @@ import {
   FilterConfiguration,
   NativeFilterType,
   Divider,
-  t,
   NativeFilterTarget,
+  logging,
 } from '@superset-ui/core';
+import { isEqual } from 'lodash';
 import { DASHBOARD_ROOT_ID } from 'src/dashboard/util/constants';
-import {
-  FilterRemoval,
-  NativeFiltersForm,
-  FilterHierarchy,
-  FilterHierarchyNode,
-} from './types';
+import { FilterRemoval, NativeFiltersForm } from './types';
 
 export const REMOVAL_DELAY_SECS = 5;
 
+export const hasCircularDependency = (
+  form: FormInstance<NativeFiltersForm>,
+  filterConfigMap: Record<string, Filter | Divider>,
+  filterId: string,
+  trace: string[] = [],
+) => {
+  if (trace.includes(filterId)) {
+    return true;
+  }
+  const formItem = form.getFieldValue('filters')?.[filterId];
+  const configItem = filterConfigMap[filterId];
+  const dependencies: string[] = formItem
+    ? 'dependencies' in formItem && formItem.dependencies
+    : 'cascadeParentIds' in configItem && configItem?.cascadeParentIds;
+
+  if (dependencies) {
+    let result = false;
+    dependencies.forEach(dependency => {
+      result ||= hasCircularDependency(form, filterConfigMap, dependency, [
```

Comment:
Maybe we should return once result is true instead of checking all dependencies?
Like:
```
if (dependencies) {
    return dependencies.some(dependency => hasCircularDependency(form, filterConfigMap, dependency, [
        ...trace,
        filterId,
      ]));
}